File: shared/src/commonMain/kotlin/com/movieskmp/compose/movieskmpcompose/base/mvvm/Navigation/NavRegistrar.kt
package com.base.mvvm.Navigation

import com.base.mvvm.ViewModels.PageViewModel
import org.koin.core.component.KoinComponent
import org.koin.core.component.get
import org.koin.core.module.Module
import org.koin.dsl.module
import kotlin.experimental.ExperimentalObjCName
import kotlin.native.ObjCName

//This class used to map ViewModel to Page.
//Currently, this class only used by Kotlin. Swift side has its own implementation.
@OptIn(ExperimentalObjCName::class)
@ObjCName(name = "DO_NOT_USE_ON_SWIFT_NavRegistrar", exact = true) //Do not use this class on swift
class NavRegistrar : KoinComponent
{
    public val navPages = mutableListOf<NavPageInfo>()
    public val modules: MutableList<Module> = mutableListOf()

    /**
     * Registers a ViewModel–Page pair for navigation.
     *
     * @param createPage factory lambda to create a page
     * @param createViewModel factory lambda to create its ViewModel
     */
    inline fun <reified TViewModel : PageViewModel, reified TPage : IPage> RegisterPageForNavigation(noinline createPage: () -> TPage, noinline createViewModel: () -> TViewModel)
    {
        // Dynamically register page and viewModel in Koin
        val pages = module()
        {
            factory<TPage> { createPage() }
            factory<TViewModel> { createViewModel() }
        }
        modules += pages;


        val vmName = TViewModel::class.simpleName ?: throw IllegalArgumentException("Cannot get ViewModel name for ${TViewModel::class}")

        // register in navPages
        if (navPages.none { it.vmName == vmName })
        {
            navPages.add(NavPageInfo(vmName = vmName,
                                     createPageFactory = { get<TPage>() },
                                     createVmFactory = { get<TViewModel>() }))
        }
    }

    inline fun <reified TViewModel : PageViewModel> CreatePage(parameters: INavigationParameters): IPage
    {
        val vmName = TViewModel::class.simpleName
            ?: throw IllegalArgumentException("ViewModel class has no simple name.")
        return CreatePage(vmName, parameters)
    }

    /**
     * Creates a page and binds its ViewModel using the stored registration.
     */
    fun CreatePage(vmName: String, parameters: INavigationParameters): IPage
    {
        val pageInfo = navPages.firstOrNull { it.vmName == vmName }

        if(pageInfo == null)
        {
            throw IllegalArgumentException("ViewModel '$vmName' was not registered for navigation.")
        }

        val page = pageInfo.createPageFactory()
        val vm = pageInfo.createVmFactory()

        page.ViewModel = vm
        vm.Initialize(parameters)

        return page
    }
}


//**********************************USAGE***************************************

//val registrar = NavRegistrar()
//val appModule = module {
//    single { registrar } // register as a instance
      //other app services
//    single<IMyService> { MyServiceImpl() }
//    single<IDbTransferService> { DbTransferService() }
//}
//registrar.registerPageForNavigation(pageFactory = { MyPage() },viewModelFactory = { MyPageViewModel() })
//startKoin {
//    modules(appModule, registrar.module)
//}

//use CreatePage() in services
//val registrar: NavRegistrar by inject()
///registrar.CreatePage<TViewModel>(parameters)

File: iosApp/iosApp/Services/Mvvm/NavRegistrar.swift
import SharedAppCore

import Foundation


class NavRegistrar {
    
    private(set) static var navPages: [NavPageInfo] = []
    
    // Registers a ViewModel–Page pair for navigation
    static func RegisterPageForNavigation<TViewModel: PageViewModel, TPage: IPage>(_
        createPage: @escaping () -> TPage,
        _ createViewModel: @escaping () -> TViewModel
    )
    {
        // Use the type name as identifier
        let vmName = String(describing: TViewModel.self)
        
        // Register into navPages only once per VM type
        if !navPages.contains(where: { $0.vmName == vmName })
        {
            let pageInfo = NavPageInfo(
                vmName: vmName,
                createPageFactory: { createPage() },
                createVmFactory: { createViewModel() }
            )
            
            navPages.append(pageInfo)
        }
    }
    
    // Generic overload for creating by type
    static func CreatePage<TViewModel: PageViewModel>(_ type: TViewModel.Type, parameters: INavigationParameters) -> IPage
    {
        let vmName = String(describing: TViewModel.self)
        return CreatePage(vmName: vmName, parameters: parameters)
    }
    
    // Creates a page and binds its ViewModel using stored registration.
    static func CreatePage(vmName: String, parameters: INavigationParameters) -> IPage
    {        
        guard let pageInfo = navPages.first(where: { $0.vmName == vmName }) else {
            fatalError("ViewModel '\(vmName)' was not registered for navigation.")
        }
        
        let page = pageInfo.createPageFactory()
        let vm = pageInfo.createVmFactory()
        
        page.ViewModel = vm        
        
        vm.Initialize(parameters: parameters)
        
        return page
    }
}


